### PR TITLE
Data aug

### LINKED ISSUE #26: Promix Naive
**Relevant background**

We assume that 5-15% of the marker positive/negative labels in our dataset are wrong. To cope with this, we will adapt the clean sample selection procedure from [ProMix](https://arxiv.org/pdf/2207.10276.pdf) from the image-level to the cell-level. 

**Design overview**

We adapt the methods presented as ProMix Naive for the binary pixel-wise classification case as follows:
1. Class-wise small-loss selection:
- after each epoch they split the training data into subsets based on its training label, s.t. one has C subsets S for each label c_0,c_1,..., c_C. They calculate the CE loss for the samples in each subset and select for each subset the top-k samples with the lowest loss as their clean training data, thus arriving at a roughly balanced sample
- We'll calculate the running 5%-percentile of the loss on foreground GT areas, separately for GT positive and negative areas. Then we'll mask out the loss where it is higher than the running 5% percentile of each class

2. Matched High-Confidence Selection
- they calculate confidence scores for each training sample and sample the ones where the confidence is higher than a given threshold tau (tau = 0.99 for CIFAR10N and tau = 0.95 for CIFAR100N)
- We mask out areas in predictions where the predicted confidence is lower than the threshold tau = 0.95

The loss and confidence scores used in the above calculations are based on **un-augmented** training data

3. Consistency Regularization
- they augment the images with weak augmentations (zoom/crop, color jitter, gaussian blur) and strong augmentations (AutoContrast, Equalize, Rotate, Solarize, Color, Posterize, Contrast, Brightness, Sharpness, Shear, Translate) and calculate the loss based on the predictions and the GT. So this is **not augmentation consistency training** but just normal data augmentation. 
- I'll just keep my augmentation pipeline as is, since I don't think its much different

4. Mixup
- they use mixup with sigma~Beta(4,4) 
- we'll just use the [segmentation version of mixup](https://openreview.net/pdf?id=rkBBChjiG)

**Code mockup**

For 1. and 2. we'll sub-class `ModelBuilder` and change class functions 
- `prep_data`: exclude mapping the data augmentation onto the train dataset
- `prep_model`: exclude `model.compile` from here
- `train`:  write custom training loop that looks like the one below

add class functions
- `class_wise_loss_selection(y_pred, y_gt)`:  calculates k-th percentile of GT positive and negative CE loss per class and stores them via exponential-moving-average in a class attribute, returns a loss mask based on this 
- `matched_high_confidence_selection(y_pred, y_gt)`: returns a loss mask for all predictions whose confidence are below a certain threshold tau

```
for x,y_gt in data:
    y_pred = model(x)
    loss_mask = model.class_wise_loss_selection(y_pred, y_gt)
    loss_mask += model.matched_high_confidence_selection(y_pred, y_gt)
    x_aug, loss_mask_aug, y_gt_aug = aug_fn(x, loss_mask, y_gt)
    with tf.gradient_tape as tape:
        y_pred_aug = model(x_aug)
        loss_value = model.loss_fn(y_gt_aug, y_pred_aug)
        loss_value = model.masked_loss_fn(loss_value, loss_mask_aug)

    grads = tape.gradient(loss_value, model.trainable_weights)
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
```
Code inside the training loop should be put into a static class function decorated by `@tf.function` in order to use static-graph mode during training. Probably we need to change the augmentation library we're using, this could be a good time to switch to one of the tf-based augmentation libraries.

**Required inputs**

Provided a description of the required inputs for this project, including column names for dfs, dimensions for image data, prompts for user input, directory structure for loading data, etc

**Output files**

Provide a description of the outputs for this project. If any plots will be generated, provide (simple) sketches demonstrating the plot type and axes labels. 

**Timeline**
Give a rough estimate for how long you think the project will take. In general, it's better to be too conservative rather than too optimistic. 
- [X] A couple days
- [ ] A week
- [ ] Multiple weeks. For large projects, make sure to agree on a plan that isn't just a single monster PR at the end. 

Estimated date when a fully implemented version will be ready for review:

Estimated date when the finalized project will be merged in:

### PR DESCRIPTION
**Pull request by JLrumberger**
**What is the purpose of this PR?**

This PR closes #26 by adding the tf augmentation pipeline and mixup.

**How did you implement your changes**

Added classes `Flip`, `Rot90`, `GaussianNoise`, `GaussianBlur`, `Zoom`, `LinearContrast`, `MixUp` and `Augmenter`. Augmenter applies augmentations on each image in a batch individually. MixUp works directly on batches, thus shouldn't be passed to Augmenter.

**Remaining issues**

None

### CODE REVIEW

File: cell_classification/augmentation_pipeline.py [hunk 1]
```diff
@@ -137,3 +138,351 @@ def py_aug(batch, tf_aug):
...
+        if tf.random.uniform(()) < self.prob:
+            zoom_factor = tf.random.uniform([1], self.min_zoom, self.max_zoom)
+            zoom_factor = 1/zoom_factor
+            zooms = tf.expand_dims(
+                tf.cast(
+                    tf.concat([zoom_factor, zoom_factor], axis=0), dtype=tf.float32
+                ), 0
+            )
+            squeeze = False
+            if tf.rank(image) == 3:
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py:300]: Are all of the channels currently stored as row x col x 1 3D images, 1 x row x col x 1 4D images, or does it depend on how they were saved?

File: cell_classification/augmentation_pipeline.py [hunk 2]
```diff
@@ -137,3 +138,351 @@ def py_aug(batch, tf_aug):
...
+            tf.Tensor:
+                The augmented image batch
+            tf.Tensor:
+                The augmented label batch
+        """
+        labels = tf.cast(labels, dtype=tf.float32)
+        undecided = tf.where(labels == 2, True, False)
+        b = tf.shape(image)[0]
+        prob = tf.random.uniform([b], 0, 1) < self.prob
+        beta = tf.random.uniform([b], 0, 1)
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py:397]: Some more comments here would be useful to explain how the vector `b` is being used to determine which images are augmented, the transformations to get the desired distribution, etc

File: cell_classification/augmentation_pipeline.py [hunk 3]
```diff
@@ -137,3 +138,351 @@ def py_aug(batch, tf_aug):
...
+            tf.Tensor:
+                The augmented labels
+        """
+        if tf.random.uniform(()) < self.prob:
+            factor = tf.random.uniform((), self.min_factor, self.max_factor)
+            image *= factor
+        return image, labels
+
+
+class MixUp(tf.Module):
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py:360]: Using MixUp, even without a schedule for `p` for promix, will help to avoid the issue we saw of the predictions being heavily skewed towards 0 and 1. Even with only high-confidence positives and negatives in the original training data, we'll now have targets that are in the middle regime

File: cell_classification/promix_naive.py [hunk 4]
```diff
@@ -69,11 +72,15 @@ def train_step(model, optimizer, loss_fn, aug_fn, loss_mask, x, y_gt):
         Returns:
             tf.Tensor: loss value
         """
-        x_aug, loss_mask_aug, y_gt_aug = aug_fn(x, loss_mask, y_gt)
+        loss_mask = tf.expand_dims(tf.cast(loss_mask, tf.uint8), -1)
+
+        x_aug, cat = aug_fn(x, tf.concat([loss_mask, y_gt], axis=-1))
+        loss_mask_aug, y_gt_aug = tf.split(cat, [loss_mask.shape[-1], y_gt.shape[-1]], axis=-1)
```
Comment (ngreenwald) [inline cell_classification/promix_naive.py]: Couldn't follow this logic

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks great, just some small comments. Did you already check the outputs for the whole pipeline? Can you share the images once you do? I'm curious to see what the range of augmentations is for 1) the non-mixup part of the pipeline and 2) the mixup part of the pipeline. It would be great if you could take a single image, and run it through the pipeline a bunch of different times to get a sense for what the range of modifications is. If you'd rather do that later, not part of this PR, that's totally fine

File: cell_classification/augmentation_pipeline.py (hunk 1 above)
Comment (JLrumberger) [reply cell_classification/augmentation_pipeline.py:300]: The `tf.map` function in augmenter squeezes the batch dimension. Since we'll probably use the augmentations only via `Augmenter` and thus `tf.map` I could get rid of the control flow and just assume HWC everywhere inside the augmentations.

File: cell_classification/augmentation_pipeline.py (hunk 2 above)
Comment (JLrumberger) [reply cell_classification/augmentation_pipeline.py:397]: 100%, the mixup function is not really readable atm.

**Comment by JLrumberger** [comment]
Non mixup part
image
Without augmentations, with augmentations and with MixUp
image

In the promix paper they parameterize like this Beta(4.0,4.0), thus their actual \sigma \in [0.84, 1] which is used to weight image1 and image2 in MixUp. Honestly I don't see how MixUp improves results yet, qualitatively the images just look weird with it.

File: cell_classification/promix_naive.py (hunk 4 above)
Comment (JLrumberger) [reply cell_classification/promix_naive.py]: `aug_fn` applies non-spatial augmentations (noise, blur, contrast) and bilinear interpolation for the spatial augmentations only for the first argument, for the second argument nearest neighbor interpolation is used and no non-spatial augmentations. Since we want noise/blur/contrast only on the multiplexed image, we split and concat stuff around here..

File: cell_classification/prepare_data_script.py [hunk 5]
```diff
@@ -25,7 +25,7 @@ def naming_convention(fname):
     normalization_dict_path=os.path.normpath(
        "C:/Users/lorenz/Desktop/angelo_lab/TONIC/normalization_dict.json"
     ),
-    normalization_quantile=0.99,
+    normalization_quantile=0.999,
```
Comment (ngreenwald) [inline cell_classification/prepare_data_script.py:28]: Will you plot a couple side-by-side, along with the raw image scaled 0-1 to max, to see what the effects are?

File: cell_classification/augmentation_pipeline.py [hunk 6]
```diff
@@ -130,10 +133,373 @@ def py_aug(batch, tf_aug):
...
+        beta = tf.math.maximum(beta, 1 - beta)
+        beta = tf.math.pow(beta, 1 / self.alpha)
+        beta = tf.where(prob, beta, tf.ones_like(beta))
+        prob = tf.cast(tf.reshape(prob, [b, 1, 1, 1]), tf.float32)
+        # apply mixup on data
+        beta = tf.reshape(beta, [b, 1, 1, 1])
+        x_mplex = beta * x_mplex + (1 - beta) * tf.reverse(x_mplex, axis=[0])
+        x_binary = beta * x_binary + (1 - beta) * tf.reverse(x_binary, axis=[0])
+        labels = beta * labels + (1 - beta) * tf.reverse(labels, axis=[0])
+        # loss mask should be 1 where both mixup loss masks == 1 and zero otherwise
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py:407]: Is this the new way you're handling 2s? How come both == 1 is the correct approach for loss masks?

File: cell_classification/promix_naive.py [hunk 7]
```diff
@@ -22,7 +23,42 @@ def __init__(self, params):
...
         self.loss_fn = self.prep_loss()
         self.quantile = self.params["quantile"]
+        self.quantile_start = self.params["quantile"]
+        self.quantile_end = self.params["quantile_end"]
+        self.quantile_warmup_steps = self.params["quantile_warmup_steps"]
         self.class_wise_loss_quantiles = {}
+        self.aug_fn = prepare_keras_aug(params, dtype=(tf.float32, tf.uint8))
+        self.mixup_fn = MixUp(prob=params["mixup_prob"], alpha=params["mixup_alpha"])
+
+    def quantile_scheduler(self, step):
```
Comment (ngreenwald) [inline cell_classification/promix_naive.py:32]: Nice, I'll be curious to see how this changes the output, I imagine it will be quite helpful

File: cell_classification/promix_naive_test.py [hunk 8]
```diff
@@ -72,6 +72,7 @@ def test_train():
         params["quantile"] = 0.3
         params["ema"] = 0.01
         params["confidence_thresholds"] = [0.1, 0.9]
+        params["mixup_prob"] = 0.0
```
Comment (ngreenwald) [inline cell_classification/promix_naive_test.py]: Should we add a test with mixup enabled just to make sure it runs?

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks great, just a couple questions

File: cell_classification/augmentation_pipeline.py (hunk 6 above)
Comment (JLrumberger) [reply cell_classification/augmentation_pipeline.py:407]: Would you rather do the promix weighted sum over the two loss_masks? I didn't know how to approach this and I thought the conservative solution is just masking out both areas, but then a lot of stuff could be masked out.. might be more clever to have to have it weighted then..

File: cell_classification/prepare_data_script.py (hunk 5 above)
Comment (JLrumberger) [reply cell_classification/prepare_data_script.py:28]: Will do before I create the decidua dataset.

**Review by ngreenwald — APPROVED** [review]
Looking forward to seeing the results!
